### PR TITLE
Add DepositERC20.sol, DepositNFT.sol and Batch checkpoints

### LINKED ISSUE #29: WIP: Clarify minimal requirement for checkpoint
## Overview

Clarify minimal requirement for checkpoint and optimistic mass exit.

## What problem can be solved

Can make many checkpoints and exits with lower gas cost.

## How does it work in detail

Rough design.

```
struct Exit {
    Range range;
    uint256 plasmaBlockNumber;
    bytes32 hashOfLeftData; // StateObject etc...
    uint256 depositedRangeId;
} // 96 bytes
```

### startCheckpoint

```
function startCheckpoint(
    Exit[] exitList,
) public
```
If you submit 20 checkpoints, call data size is 20*96 = 1,920 bytes.
sha3 cost is 390 gas.
SSTORE cost is 20000 gas.
cd_gas_per_byte = 68. 390 + 20000 + 130560gas, so 7548 gas per exit 

### ChallengeInclusion

```
function challengeInclusion(
    Exit[] exitList,
    Exit targetExit,
) public
```

### Predicate sides

* deprecateExit

```
function proveExitDeprecation(
    Exit[] exitList,
    Checkpoint _deprecatedExit,
    Transaction _transaction,
    bytes _witness,
    StateUpdate _postState
) public
```

1. check _deprecatedExit is included in batched exits
    * And its predicateAddress equals to seld.address
2. prove deprecation of _deprecatedExit
3. call deprecateExit(_deprecatedExit)

### PR DESCRIPTION
**Pull request by syuhei176**
## Description

Add DepositERC20.sol and DepositNFT.sol.

These contracts include only `deposit` function, don't include exit-game related functions yet.

- [x] Add DepositERC20(deposit only)
- [x] Add DepositNFT(deposit only)
- [x] Add deposit and predicates folder
- [x] Minimal batch checkpoints and exits

Close #29

### CODE REVIEW

File: contracts/standard/DepositStandard.sol [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+        Challenge memory _challenge,
+        bytes memory _challengedCheckpointInclusionProof,
+        bytes memory _challengingCheckpointInclusionProof
+    ) public {
+        bytes32 challengedCheckpointId = calculateMerkleRootWithProof(_challenge.challengedCheckpoint, _challengedCheckpointInclusionProof);
+        bytes32 challengingCheckpointId = calculateMerkleRootWithProof(_challenge.challengingCheckpoint, _challengingCheckpointInclusionProof);
+        bytes32 challengeId = getChallengeId(_challenge);
+        // Verify that the challenge may be added
+        require(challenges[challengeId], "Challenge must exist");
+        require(!exitExists(challengingCheckpointId), "Challenging exit must no longer exist");
```
Comment (syuhei176) [inline contracts/standard/DepositStandard.sol]: There are batch checkpoint 1 and batch checkpoint 2, and challenge "batch checkpoint 2" with "batch checkpoint 1". When one of the checkpoints(called "Checkpoint D") in "batch checkpoint 1" is deprecated, "batch checkpoint 2"'s `outstandingChallenges` will be decremented even if "Checkpoint D" doesn't intersect with "batch checkpoint 1".

Comment (yuriko627) [reply contracts/standard/DepositStandard.sol]: what do you mean by decremented?

Comment (shogochiai) [reply contracts/standard/DepositStandard.sol]: decreased. yeah it's not verb but programmers can guess it.

outstanding means "current number". and it implied some integer is kept in there for coubting something for security

Comment (shogochiai) [reply contracts/standard/DepositStandard.sol]: 1. Every checkpoint is defined as batch checkpoints

2. Invalidation classification
    2-1) Challenges a checkpoint by a checkpoint - invalidates whole checkpoint
    2-2) Deprecates a checkpoint by a tx  - invalidates only a deprecated range

3. The checkpoint initiator can assume that his planned checkpoint is correct or not, by watching history. But challenge-for-checkpoint is unpredictale for initiator.

4. Example
    4-1) A group of people initiate a batch checkpoint
    4-2) One of them double spend his State Update for other person
    4-3) The whole batch checkpoint would be invalidated

5. Side-note
    5-1. Checkpoint requires a commit-of-plasma-contract
    5-2. So, before checkpoint creation, initiator can confirm the history of all txs

**Pull request closed without merging**